### PR TITLE
Conditionally create server spans for flask

### LINKED ISSUE #446: Flask: Conditionally create SERVER spans
Part of #445

### PR DESCRIPTION
**Pull request by ashu658**
# Description

Adds support to make spans as INTERNAL if a span is already present in current context in flask. 

Fixes #446 

## Type of change

Please delete options that are not relevant.

- [x] Breaking change (fix or feature that would cause existing functionality to not work as expected)
Currently flask only makes SERVER spans. With this PR it can make INTERNAL spans in presence of a span in current context.

# How Has This Been Tested?

Added unit test for the changes.
Manually tested the scenario.
To reproduce the issue: Add two instrumentors i.e. FlaskInstrumentor and  OpenTelemetryMiddleware in the same app. This creates two separate traces instead of one.

# Does This PR Require a Core Repo Change?

- [x] No.

# Checklist:

See [contributing.md](https://github.com/open-telemetry/opentelemetry-python-contrib/blob/main/CONTRIBUTING.md) for styleguide, changelog guidelines, and more.

- [x] Followed the style guidelines of this project
- [x] Changelogs have been updated
- [x] Unit tests have been added
- [x] Documentation has been updated

### CODE REVIEW

File: instrumentation/opentelemetry-instrumentation-flask/src/opentelemetry/instrumentation/flask/__init__.py [hunk 1]
```diff
@@ -229,7 +237,9 @@ def _teardown_request(exc):
             activation.__exit__(
                 type(exc), exc, getattr(exc, "__traceback__", None)
             )
-        context.detach(flask.request.environ.get(_ENVIRON_TOKEN))
+
+        if flask.request.environ.get(_ENVIRON_TOKEN, None) is not None:
```
Comment (lzchen) [inline instrumentation/opentelemetry-instrumentation-flask/src/opentelemetry/instrumentation/flask/__init__.py:241]:
```suggestion
        if flask.request.environ.get(_ENVIRON_TOKEN):
```

File: instrumentation/opentelemetry-instrumentation-flask/src/opentelemetry/instrumentation/flask/__init__.py [hunk 2]
```diff
@@ -176,15 +176,22 @@ def _before_request():
...
-            extract(flask_request_environ, getter=otel_wsgi.wsgi_getter)
-        )
+
+        token = ctx = None
+        span_kind = trace.SpanKind.INTERNAL
+
+        if trace.get_current_span() is trace.INVALID_SPAN:
+            ctx = extract(flask_request_environ, getter=otel_wsgi.wsgi_getter)
+            token = context.attach(ctx)
+            span_kind = trace.SpanKind.SERVER
```
Comment (lzchen) [inline instrumentation/opentelemetry-instrumentation-flask/src/opentelemetry/instrumentation/flask/__init__.py:187]: If span is found in current context, remote span is assume to be existing, so I believe you'll have to set the current context as the parent.

File: instrumentation/opentelemetry-instrumentation-flask/src/opentelemetry/instrumentation/flask/__init__.py [hunk 3]
```diff
@@ -176,15 +176,23 @@ def _before_request():
...
-        token = context.attach(
-            extract(flask_request_environ, getter=otel_wsgi.wsgi_getter)
-        )
+
+        token = ctx = None
+        span_kind = trace.SpanKind.INTERNAL
+
+        if trace.get_current_span() is trace.INVALID_SPAN:
+            ctx = extract(flask_request_environ, getter=otel_wsgi.wsgi_getter)
+            print(ctx)
```
[the remark this answers is not in the document]
Comment (ashu658) [reply instrumentation/opentelemetry-instrumentation-flask/src/opentelemetry/instrumentation/flask/__init__.py]: removed.

File: instrumentation/opentelemetry-instrumentation-flask/src/opentelemetry/instrumentation/flask/__init__.py (hunk 1 above)
Comment (ashu658) [reply instrumentation/opentelemetry-instrumentation-flask/src/opentelemetry/instrumentation/flask/__init__.py]: done.

File: instrumentation/opentelemetry-instrumentation-flask/src/opentelemetry/instrumentation/flask/__init__.py (hunk 2 above)
Comment (ashu658) [reply instrumentation/opentelemetry-instrumentation-flask/src/opentelemetry/instrumentation/flask/__init__.py:189]: Done.
I think it was passing earlier as `tracer.start_span()` makes current span as parent if context passed is `none`. 
@lzchen Should we set parent explicitly or let `start_span()` handle it?

Comment (lzchen) [reply instrumentation/opentelemetry-instrumentation-flask/src/opentelemetry/instrumentation/flask/__init__.py:189]: Oh yes you are correct. Either way is fine with me.

File: instrumentation/opentelemetry-instrumentation-django/tests/test_middleware_asgi.py [hunk 4]
```diff
@@ -59,7 +59,7 @@
 if DJANGO_2_0:
     from django.urls import re_path
 else:
-    from django.conf.urls import url as re_path
+    from django.conf.urls import url as re_path  # pylint: disable=E0611
```
Comment (ashu658) [inline instrumentation/opentelemetry-instrumentation-django/tests/test_middleware_asgi.py:62]: Added this as django.conf.urls.url() is removed in django 4.0. leading to lint check failure.

**Comment by owais** [comment]
The title says flask but contains code for Django.

**Comment by ashu658** [comment]
> The title says flask but contains code for Django.

**Comment by ashu658** [comment]
> The title says flask but contains code for Django.

Just saw these changes have already been merged to master. Removed the django related code.

**Review by owais — APPROVED** [review]

**Pull request merged**